blib/man3/Debug.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Debug 3"
.TH Debug 3 "2008-02-19" "perl v5.8.8" "User Contributed Perl Documentation"
.SH "NAME"
Debug \- Perl extension for logging/debugging.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  use Debug;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Debug provide a basic framework for the logging/debugging of Perl modules.
.PP
\&\fInew()\fR
Create a new Debug object
.PP
\&\fItimeStamp()\fR
Enables date and timestamping
.PP
\&\fIdetail()\fR
Enables detailed logging
.PP
debug
Enables debugging
.PP
debugFile
Set a debug file for logging to a file
.PP
initialize
Initializes debugging/logging
.PP
message	
Write a message to configured destinations
.PP
my \f(CW$debug\fR = new \fIDebug()\fR;
\&\f(CW$debug\fR\->debugFile(\*(L"debug.log\*(R");
\&\f(CW$debug\fR\->\fIinitialize()\fR;
\&\f(CW$debug\fR\->message(\*(L"Test 1\*(R");
\&\f(CW$debug\fR\->\fItimeStamp()\fR;
\&\f(CW$debug\fR\->message(\*(L"Test 2\*(R");
\&\f(CW$debug\fR\->\fIdetail()\fR;
\&\f(CW$debug\fR\->message(\*(L"Test 3\*(R");
.Sh "\s-1EXPORT\s0"
.IX Subsection "EXPORT"
None by default.
.SH "AUTHOR"
.IX Header "AUTHOR"
\&\s-1JJMB\s0, <lt>jjmb@jjmb.com>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
perl.


Changes
Revision history for Perl extension Debug.

0.01  Fri Jun 14 05:53:34 2002
	- original version; created by h2xs 1.21 with options
		-AXcn Debug



Debug.pm
package Debug;

use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Debug ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);
our $VERSION = '0.02';


# Preloaded methods go here.
my $package = __PACKAGE__;
my $dfh = "";
local *DFH;

# new() - creates a new Debug object.
sub new {
	my $class = shift;
	my $debug = shift || 1;
	my $self = {};
	
	$self->{'DEBUG'} = $debug if defined $debug;

	bless $self, $class;
	return $self;
}

# timeStamp() - enables date and timestamped messages
sub timeStamp() {
	my $self = shift;
	my $timestamp = shift || 1;
	
	$self->{'TIMESTAMP'} = $timestamp if defined $timestamp;
	
	return $self->{'TIMESTAMP'};
}

# detail() - enables detailed messages including line of code and callers
sub detail() {
	my $self = shift;
	my $detail = shift || 1;
	
	$self->{'DETAIL'} = $detail if defined $detail;
	
	return $self->{'DETAIL'};
}

# debug() - enabled normal messages
sub debug {
	my $self = shift;
	my $debug = shift;
	
	$self->{'DEBUG'} = $debug if defined $debug;
	
	return $self->{'DEBUG'};
}

# debugFile() - sets debug file 
sub debugFile {
	my $self = shift;
	my $debugfile = shift;
	
	$self->{'DEBUGFILE'} = $debugfile if defined $debugfile;
	
	return $self->{'DEBUGFILE'};
}

# initialize() - initialize logging
sub initialize {
	my $self = shift;

	if (($self->{'DEBUG'}) && ($self->{'DEBUGFILE'}))
	{
		open(FH,"+>> $self->{'DEBUGFILE'}") || die print STDERR "error:($package):could not open $self->{'DEBUGFILE'}.\n";
		*DFH = *FH;
		print STDERR "Logging to $self->{'DEBUGFILE'}\n";
	}
	elsif (($self->{'DEBUG'}) && (!defined($self->{'DEBUGFILE'})))
	{
		open(FH,">&STDOUT") || die print STDERR "error:($package):could not open STDERR.\n";
		*DFH = *FH;
		print STDERR "Logging to STDOUT\n";
	}
	else
	{
		open(FH,"+>> /dev/null") || die print STDERR "error:($package):could not open /dev/null.\n";
		*DFH = *FH;
		print STDERR "Logging to /dev/null\n";
	}

	select((select(DFH),$| = 1)[0]);

	if (*DFH)
	{
		print DFH "debug:($package): initialized.\n" || 
		die print STDERR "error:($package):could not write to ".$self->debugfilehandle().".\n";
	}
	else
	{
		print STDERR "error:($package) debug file handle does not exist.\n";
	}

	return $self;
}

# message() - writes message
sub message {
	my $self = shift;
	# my $message = shift || "null";
	my $message = shift;
	my $datetime = undef;
	my $origin = undef;
	my $msg = undef;
	
	if($message eq "") {
		$message = "null";
	}

	if($self->{'TIMESTAMP'}) {
		my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = gmtime(time);
		$datetime = sprintf("%04d-%02d-%02d.%02d:%02d:%02d",$year+1900,$mon+1,$mday,$hour,$min,$sec);
		
		$msg .= "[".$datetime."]:";
		# $msg .= $message;
		# print DFH "[".$datetime."]".$message."\n";
	}
	
	if($self->{'DETAIL'}) {
		# my $origin = (caller(0))[0]."->".(caller(0))[2]."->".(caller(0))[3];
		# my $origin = (caller(0))[0]."->".(caller(0))[3];
		$origin = (caller(0))[0];
		
		# Uncomment to add file name back to debug output
		# $origin .= "->".(caller())[1];
		
		# Line number
		$origin .= "(".(caller())[2].")";
		
		# Called module and function
		$origin .= "->".(caller(0))[3];
		
		$msg .= "[".$origin."]:";
		# print DFH "[".$origin."]:".$message."\n";
	}
	
	$msg .= $message;

	# print DFH "[".$datetime."]".":[".$origin."]:".$message."\n";
	print DFH $msg."\n";
}

1;
__END__
# Below is stub documentation for your module.

=head1 NAME

Debug - Perl extension for logging/debugging.

=head1 SYNOPSIS

  use Debug;

=head1 DESCRIPTION

Debug provide a basic framework for the logging/debugging of Perl modules.

new()
Create a new Debug object

timeStamp()
Enables date and timestamping

detail()
Enables detailed logging

debug
Enables debugging

debugFile
Set a debug file for logging to a file

initialize
Initializes debugging/logging

message	
Write a message to configured destinations

my $debug = new Debug();
$debug->debugFile("debug.log");
$debug->initialize();
$debug->message("Test 1");
$debug->timeStamp();
$debug->message("Test 2");
$debug->detail();
$debug->message("Test 3");

=head2 EXPORT

None by default.

=head1 AUTHOR

JJMB, <lt>jjmb@jjmb.comE<gt>

=head1 SEE ALSO

L<perl>.

=cut
